test/test_host.go

```diff
@@ -290,3 +297,95 @@ func (s *HostSuite) TestResourceLimits(t *c.C) {
 
 	assertResourceLimits(t, out.String())
 }
+
+func (s *HostSuite) TestUpdate(t *c.C) {
+	dir := t.MkDir()
+	flynnHost := filepath.Join(dir, "flynn-host")
+	run(t, osexec.Command("cp", args.FlynnHost, flynnHost))
+
+	// start flynn-host
+	id := random.String(8)
+	var out bytes.Buffer
+	cmd := osexec.Command(
+		flynnHost,
+		"daemon",
+		"--http-port", "11113",
+		"--state", filepath.Join(dir, "host-state.bolt"),
+		"--id", id,
+		"--backend", "mock",
+		"--vol-provider", "mock",
+		"--volpath", filepath.Join(dir, "volumes"),
+	)
+	cmd.Stdout = &out
+	cmd.Stderr = &out
+	defer func() {
+		debug(t, "*** flynn-host output ***")
+		debug(t, out.String())
+		debug(t, "*************************")
+	}()
+	t.Assert(cmd.Start(), c.IsNil)
+	defer cmd.Process.Kill()
+
+	httpClient := &http.Client{Transport: &http.Transport{Dial: dialer.Retry.Dial}}
+	client := cluster.NewHost(id, "http://127.0.0.1:11113", httpClient)
+
+	// exec a program which exits straight away
+	_, err := client.Update("/bin/true")
+	t.Assert(err, c.NotNil)
+	status, err := client.GetStatus()
+	t.Assert(err, c.IsNil)
+	t.Assert(status.ID, c.Equals, id)
+	t.Assert(status.PID, c.Equals, cmd.Process.Pid)
+
+	// exec a program which reads the control socket but then exits
+	_, err = client.Update("/bin/bash", "-c", "<&4; exit")
+	t.Assert(err, c.NotNil)
+	status, err = client.GetStatus()
+	t.Assert(err, c.IsNil)
+	t.Assert(status.ID, c.Equals, id)
+	t.Assert(status.PID, c.Equals, cmd.Process.Pid)
+
+	// exec flynn-host and check we get the status from the new daemon
+	pid, err := client.Update(
+		flynnHost,
+		"daemon",
+		"--http-port", "11113",
+		"--state", filepath.Join(dir, "host-state.bolt"),
+		"--id", id,
+		"--backend", "mock",
+		"--vol-provider", "mock",
+		"--volpath", filepath.Join(dir, "volumes"),
+	)
+	t.Assert(err, c.IsNil)
+	defer syscall.Kill(pid, syscall.SIGKILL)
+
+	done := make(chan struct{})
+	go func() {
+		cmd.Process.Signal(syscall.SIGTERM)
+		syscall.Wait4(cmd.Process.Pid, nil, 0, nil)
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(15 * time.Second):
+		t.Fatal("timed out waiting for flynn-host daemon to exit")
+	}
+
+	// client.GetStatus intermittently returns io.EOF right after the update. We
```

Comment:
Copying the listener over just copies the listener, all established connections will be closed when the old process exits.